File: AoC/2025_pyhs/round3/solution1.hs
module Main where

import Data.List
import Data.Char
import System.IO
import Distribution.Compat.CharParsing (digit)
import Distribution.Simple.Utils (xargs)


readInput :: String -> IO [String]
readInput fname = do
    content <- readFile fname
    return $ lines content

printInput :: [String] -> IO ()
printInput input = mapM_ putStrLn input

stringToInt :: String -> [Int]
stringToInt xs = [digitToInt x | x <- xs] -- has to be digits only

whereFirst :: [Int] -> Int -> Int -> Int -- Maybe Int
whereFirst (x:xs) target index = case x == target of True  -> index
                                                     False -> whereFirst xs target (index + 1)

whereMax :: [Int] -> Int
whereMax xs = whereFirst xs (maximum xs) 0

dropLast :: [a] -> [a]
dropLast xs = take (length xs - 1) xs

takeAfter :: Int -> [a] -> [a]
takeAfter ind xs = drop (ind + 1) xs -- guaranteed to be in bounds

getFstIndex :: String -> Int
getFstIndex s = whereMax $ stringToInt (dropLast s)

getJoltage :: String -> Int
getJoltage s = read joltage :: Int
    where ind = getFstIndex s
          digit1 = s !! ind : ""
          rest = [digitToInt c | c <- takeAfter ind s]
          digit2 = show $ maximum rest
          joltage = digit1 ++ digit2


main :: IO ()
main = do
    input <- readInput "input.txt"
    -- printInput input
    let total = sum $ map getJoltage input
    print total

File: AoC/2025_pyhs/round3/solution1.py

import numpy as np

def read_input(file_path):
    result = []
    with open(file_path, "r") as f:
        for line in f.readlines():
            result.append([int(i) for i in line.strip()])
    return np.array(result)

def get_joltage(x):
    head = x[:-1]
    fst_digit = head.max()
    fst_index = np.where(x == fst_digit)[0][0]
    snd_digit = x[fst_index+1:].max()
    return int(str(fst_digit) + str(snd_digit))
    

if __name__ == "__main__":
    fname = "input.txt"
    data = read_input(fname)
    voltages = map(get_joltage, data)
    print(sum(voltages))
